File: server-csharp/Magician/Tables.cs
using System.Numerics;
using SpacetimeDB;

public static partial class Module
{
    [Table(Name = "magician", Public = true)]
    [SpacetimeDB.Index.BTree(Name = "SameMatchPlayers", Columns = [nameof(MatchId), nameof(Id)])]
    public partial struct Magician
    {
        [PrimaryKey]
        public Identity identity;

        [Unique, AutoInc]
        public uint Id;
        public string Name;

        [SpacetimeDB.Index.BTree(Name = "MatchId")]
        public uint MatchId;
        public DbVector3 Position;
        public DbRotation2 Rotation;
        public DbVector3 Velocity;
        public DbVector3 CorrectedVelocity;
        public ComplexCollider Collider;
        public List<CollisionEntry> CollisionEntries;
        public bool IsColliding;
        public MagicianState State;
        public KinematicInformation KinematicInformation;
        public List<PermissionEntry> PlayerPermissionConfig;
        public List<Timer> Timers;

        //public List<ThrowingCard> AttackMagazine;
    }
    
   [Table(Name = "move_all_magicians", Scheduled = nameof(MoveMagicians), ScheduledAt = nameof(scheduled_at))]
    public partial struct Move_All_Magicians_Timer
    {
        [PrimaryKey, AutoInc] public ulong scheduled_id;
        public ScheduleAt scheduled_at;
        public float tick_rate;
    }

    [Table(Name = "gravity_magician", Scheduled = nameof(ApplyGravityMagician), ScheduledAt = nameof(scheduled_at))]
    public partial struct Gravity_Timer_Magician
    {
        [PrimaryKey, AutoInc] public ulong scheduled_id;
        public ScheduleAt scheduled_at;
        public float tick_rate;
        public float gravity;
    }

    [Table(Name = "handle_magician_timers_timer", Scheduled = nameof(HandleMagicianTimers), ScheduledAt = nameof(scheduled_at))]
    public partial struct Handle_Magician_Timers_Timer
    {
        [PrimaryKey, AutoInc] public ulong scheduled_id;
        public ScheduleAt scheduled_at;
        public float tick_rate;
    }

    
}

File: server-rust/spacetimedb/src/Magician/Tables.rs
use std::time::Duration;
use spacetimedb::{table, rand::Rng, Identity, SpacetimeType, ReducerContext, ScheduleAt, Table, Timestamp};
use crate::*;


#[spacetimedb::table(name = magician, public, index(name = same_game_players, btree = [game_id, id]), index(name = game_id, btree = [game_id]))]
pub struct Magician {
    #[primary_key] pub identity: Identity,
    #[unique] #[auto_inc] pub id: u32,
    pub name: String,
    pub game_id: u32,
    pub position: DbVector3,
    pub rotation: DbRotation2,
    pub velocity: DbVector3,
    pub corrected_velocity: DbVector3,
    pub collider: ComplexCollider,
    pub collision_entries: Vec<CollisionEntry>,
    pub is_colliding: bool,
    pub state: MagicianState,
    pub kinematic_information: KinematicInformation,
    pub player_permission_config: Vec<PermissionEntry>,
    pub timers: Vec<Timer>,
    pub bullets: Vec<ThrowingCard>,
    pub bullet_capacity: i32,
}

#[spacetimedb::table(name = move_all_magicians, scheduled(move_magicians))]
pub struct MoveAllMagiciansTimer {
    #[primary_key] #[auto_inc] pub scheduled_id: u64,
    pub scheduled_at: ScheduleAt,
    pub tick_rate: f32,
    pub game_id: u32,
}

#[spacetimedb::table(name = gravity_magician, scheduled(apply_gravity_magician))]
pub struct GravityTimerMagician {
    #[primary_key] #[auto_inc] pub scheduled_id: u64,
    pub scheduled_at: ScheduleAt,
    pub tick_rate: f32,
    pub gravity: f32,
    pub game_id: u32,
}

#[spacetimedb::table(name = handle_magician_timers_timer, scheduled(handle_magician_timers))]
pub struct HandleMagicianTimersTimer {
    #[primary_key] #[auto_inc] pub scheduled_id: u64,
    pub scheduled_at: ScheduleAt,
    pub tick_rate: f32,
    pub game_id: u32,
}
